# TUGAS PRAKTIKUM 1

**Load Data**

In [ ]:
import pandas as pd
data = pd.read_csv('/content/drive/MyDrive/Machine Learning   15/Data/voice.csv')
data.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

**1. Buatlah model klasifikasi dengan menggunakan kNN untuk mengklasifikasikan jenis suara male dan female pada dataset voice.csv.**

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import pandas as pd

# Encode label column
data['label'] = data['label'].map({'male': 1, 'female': 0})

# Separate features (X) and label (y)
X = data.drop('label', axis=1)
y = data['label']

# Split data (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Standardize features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Initialize and train kNN model (using k=5 as an example)
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

# Predict on the test set
y_pred = knn.predict(X_test)

# Evaluate the model
print("Akurasi:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Laporan Klasifikasi:\n", classification_report(y_test, y_pred))

Akurasi: 0.9737118822292324
Confusion Matrix:
 [[435  17]
 [  8 491]]
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       0.98      0.96      0.97       452
           1       0.97      0.98      0.98       499

    accuracy                           0.97       951
   macro avg       0.97      0.97      0.97       951
weighted avg       0.97      0.97      0.97       951



**2. Lakukan percobaan untuk mengetahui fitur-fitur yang paling optimal untuk digunakan. Fitur apa saja yang Anda gunakan untuk mendapatkan hasil terbaik?**

In [ ]:
feature_sets = [
    ['meanfreq'],
    ['meanfreq', 'sd', 'median'],
    ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR'],
    ['sp.ent', 'sfm'],
    ['meanfun', 'minfun', 'maxfun'],
    ['meandom', 'mindom', 'maxdom', 'dfrange', 'modindx'],
    list(X.columns)
]

**Perulangan untuk setiap kumpulan fitur yang telah didefinisikan**

In [ ]:
results = []
for feature_set in feature_sets:
    print(f"Mencoba kombinasi fitur: {feature_set}")

Mencoba kombinasi fitur: ['meanfreq']
Mencoba kombinasi fitur: ['meanfreq', 'sd', 'median']
Mencoba kombinasi fitur: ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR']
Mencoba kombinasi fitur: ['sp.ent', 'sfm']
Mencoba kombinasi fitur: ['meanfun', 'minfun', 'maxfun']
Mencoba kombinasi fitur: ['meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']
Mencoba kombinasi fitur: ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR', 'skew', 'kurt', 'sp.ent', 'sfm', 'mode', 'centroid', 'meanfun', 'minfun', 'maxfun', 'meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']


**Langkah berikutnya adalah:**
1. memilih fitur yang relevan
2. membagi data
3. menstandarisasi fitur
4. melatih model kNN
5. memprediksi hasil
6. mengevaluasi akurasi model
7. menyimpan hasil percobaan

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

results = []
for feature_set in feature_sets:
    print(f"Mencoba kombinasi fitur: {feature_set}")

    # Select features
    X_subset = X[feature_set]
    y_subset = y

    # Split data (70% train, 30% test)
    X_train_subset, X_test_subset, y_train_subset, y_test_subset = train_test_split(X_subset, y_subset, test_size=0.3, random_state=42)

    # Standarisasi Fitur
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train_subset)
    X_test_scaled = scaler.transform(X_test_subset)

    # Inisialisasi dan latih model kNN
    knn = KNeighborsClassifier(n_neighbors=5)
    knn.fit(X_train_scaled, y_train_subset)

    # Predict pada data uji
    y_pred_subset = knn.predict(X_test_scaled)

    # Evaluasi Model
    accuracy = accuracy_score(y_test_subset, y_pred_subset)

    # Menyimpan hasil
    results.append({'feature_set': feature_set, 'accuracy': accuracy})


Mencoba kombinasi fitur: ['meanfreq']
Mencoba kombinasi fitur: ['meanfreq', 'sd', 'median']
Mencoba kombinasi fitur: ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR']
Mencoba kombinasi fitur: ['sp.ent', 'sfm']
Mencoba kombinasi fitur: ['meanfun', 'minfun', 'maxfun']
Mencoba kombinasi fitur: ['meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']
Mencoba kombinasi fitur: ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR', 'skew', 'kurt', 'sp.ent', 'sfm', 'mode', 'centroid', 'meanfun', 'minfun', 'maxfun', 'meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']


**Identifikasi fitur terbaik, dengan melakukan iterasi pada daftar results untuk mencari feature_set dengan nilai akurasi tinggi**

In [ ]:
best_accuracy = -1
best_feature_set = None

for result in results:
    if result['accuracy'] > best_accuracy:
        best_accuracy = result['accuracy']
        best_feature_set = result['feature_set']

print(f"Kumpulan fitur terbaik adalah: {best_feature_set}")
print(f"Akurasi tertinggi adalah: {best_accuracy}")

Kumpulan fitur terbaik adalah: ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR', 'skew', 'kurt', 'sp.ent', 'sfm', 'mode', 'centroid', 'meanfun', 'minfun', 'maxfun', 'meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']
Akurasi tertinggi adalah: 0.9737118822292324


**Menampilkan hasil terbaik dari semua kombinasi fitur yang diuji, hasil ini mencakup kumpulan fitur yang memberikan akurasi tertinggi beserta nilai akurasinya**

In [ ]:
print(f"Kumpulan fitur yang menghasilkan akurasi tertinggi adalah: {best_feature_set}")
print(f"Akurasi tertinggi yang dicapai dengan kumpulan fitur ini adalah: {best_accuracy}")

Kumpulan fitur yang menghasilkan akurasi tertinggi adalah: ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR', 'skew', 'kurt', 'sp.ent', 'sfm', 'mode', 'centroid', 'meanfun', 'minfun', 'maxfun', 'meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']
Akurasi tertinggi yang dicapai dengan kumpulan fitur ini adalah: 0.9737118822292324


**Kesimpulan Tugas 1 :** kumpulan fitur terbaik untuk model klasifikasi kNN pada dataset voice.csv dengan nilai akurasi tinggi yakni 0.9737, fitur fitur tersebut adalah ['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR', 'skew', 'kurt', 'sp.ent', 'sfm', 'mode', 'centroid', 'meanfun', 'minfun', 'maxfun', 'meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']